feedback/broker.go

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2019 TFG Co <backend@tfgco.com>
+ * Author: TFG Co <backend@tfgco.com>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package feedback
+
+import (
+	"encoding/json"
+	"sync"
+
+	gcm "github.com/topfreegames/go-gcm"
+	"github.com/topfreegames/pusher/structs"
+
+	"github.com/sideshow/apns2"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+)
+
+// Message is a struct that will decode an apns or gcm feedback message
+type Message struct {
+	From             string                 `json:"from"`
+	MessageID        string                 `json:"message_id"`
+	MessageType      string                 `json:"message_type"`
+	Error            string                 `json:"error"`
+	ErrorDescription string                 `json:"error_description"`
+	DeviceToken      string                 `json:"DeviceToken"`
+	ID               string                 `json:"id"`
+	Err              map[string]interface{} `json:"Err"`
+	Metadata         map[string]interface{} `json:"metadata"`
+	Reason           string                 `json:"reason"`
+}
+
+// Broker receives kafka messages in its InChan, unmarshal them according to the
+// platform and routes them to the correct out channel after examining their content
+type Broker struct {
+	Logger              *log.Logger
+	Config              *viper.Viper
+	InChan              *chan QueueMessage
```

Comment:
Why is InChan a pointer to a chan?

Reply:
I'll change that. I wanted to avoid a new channel allocation but since it's a reference type, just the reference is passed although we don't use the pointer explicitly.